### DISCUSSION TITLE
DP 동적계획법 - 찍먹

### DISCUSSION THREAD

**Discussion by viiviii** [accident-study-group/leetcode-study#21] [General]
### 동적계획법(DP)이란
* 두 번 이상 반복 계산되는 부분 문제들의 답을 미리 저장함으로써 속도의 향상을 꾀하는 알고리즘 설계 기법

### 중복되는 부분 문제
* 큰 의미에서 분할 정복과 같이 문제를 더 작은 문제들로 나눈다는 접근 방식

### 동적 계획법과 분할 정복 차이
* 문제를 나누는 방식에 차이가 있다
* 동적 계획법에서 어떤 부분 문제는 2개 이상의 문제를 푸는데 사용됨

<img width="816" alt="image" src="https://user-images.githubusercontent.com/75404713/189522391-568d00dc-1867-444b-918f-8b97f03a12ae.png">

* a는 각 부분 문제들이 서로 연관이 없어서, 재귀 호출을 통해 하나의 부분 문제를 한 번만 해결
* b는 나눠진 각 문제들이 같은 부분 문제에 의존하므로, 재귀를 통해 각 문제를 해결하면 중복 계산이 많아짐
* 👉 이런 문제를 해결하기 위해 고안된 알고리즘 설계 기법이 동적 계획법이다

### 용어

* `중복되는 부분 문제(overlapping subproblems)` - 두 번 이상 계산되는 부분 문제
* `캐시(cache)` - 이미 계산한 값을 저장해두는 메모리의 장소
* `메모이제이션(memorization)` - 결과를 저장하는 장소를 마련해 두고, 한번 계산한 값을 저장해 뒀다 재활용하는 최적화 기법

### 메모이제이션을 적용할 수 있는 경우
- 메모이제이션은 `참조적 투명 함수`의 경우에만 적용할 수 있다
  - 결과가 외부 요소의 영향을 받는다면 캐싱을 할 수가 없으니까
- `참조적 투명성 함수(referential transparency)` - 입력이 고정되어 있을 때 결과가 항상 같은  함수

### 메모이제이션 구현 패턴
- 한 가지 패턴을 정해두고 항상 같은 형태로 구현하기로 하면 작성 및 버그 찾기가 쉽다
  - 책에서는 기저 사례를 제일 먼저 처리하는 등 항상 정해놓은 한가지 형태로 구현함

### 동적 계획법 구현
1. 주어진 문제를 완전 탐색을 이용해 해결
2. 중복된 부분 문제를 한 번만 계산하도록 메모이제이션을 적용

+) 이 방법은 대부분의 방법을 단순화한 것이고, 이 외의 구현 방법 및 유의점은 차차 친해지기..(???)

---
### 출처
- 책 <알고리즘 문제 해결전략> p.207~218

**Comment by donald-deok** [comment]
아니 선생님 종만북도 있으십니까?

**Reply by viiviii** [reply]
코테 고수를 아주 예전부터 꿈꿨지만... 이번에 펴봤습니다 흑흑

**Reply by viiviii** [reply]
흐드드님께 남겨주신 덕생님의 DP 팁 잘 읽었습니다요
